subsys/bootloader/bl_storage/bl_storage.c

```diff
@@ -72,15 +73,39 @@ uint32_t num_public_keys_read(void)
 }
 
 /* Value written to the invalidation token when invalidating an entry. */
-#define INVALID_VAL 0xFFFF0000
+#define INVALID_VAL 0x50FA0000
+#define INVALID_WRITE_VAL 0xFFFF0000
+#define VALID_VAL 0x50FAFFFF
+#define MAX_NUM_PUBLIC_KEYS 16
+#define TOKEN_IDX_OFFSET 24
 
 static bool key_is_valid(uint32_t key_idx)
 {
-	return nrfx_nvmc_uicr_word_read(&BL_STORAGE->key_data[key_idx].valid) != INVALID_VAL;
+	if (key_idx >= num_public_keys_read()) {
+		return false;
+	}
+
+	const uint32_t token = nrfx_nvmc_uicr_word_read(&BL_STORAGE->key_data[key_idx].valid);
+	const uint32_t invalid_val = INVALID_VAL | key_idx << TOKEN_IDX_OFFSET;
+	const uint32_t valid_val = VALID_VAL | key_idx << TOKEN_IDX_OFFSET;
+
+	if (token == invalid_val) {
+		return false;
+	} else if (token == valid_val) {
+		return true;
+	}
+
+	/* Invalid value. */
+	k_panic();
+	return false;
 }
 
 int verify_public_keys(void)
 {
+	if (num_public_keys_read() >= MAX_NUM_PUBLIC_KEYS) {
```

Comment:
This was correct in the earlier version.
In this case it should be simply `>`. This is a number of keys that is checked here not an index (like in line 84).